### PR TITLE
Update ReTyper to Ycheck patterns & inline fixes

### PR DESCRIPTION
**Pull request by OlivierBlanvillain**
This PR does minor adjustments to Typer and ReTyper such that patterns pass Ycheck. This is required to be able to inline pattern matching expressions.

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/ast/TreeInfo.scala [hunk 1]
```diff
@@ -512,8 +512,8 @@ trait TypedTreeInfo extends TreeInfo[Type] { self: Trees.Instance[Type] =>
       val (meth, targs, argss) = decomposeCall(fn)
       (meth, targs, argss :+ args)
     case TypeApply(fn, targs) =>
-      val (meth, Nil, Nil) = decomposeCall(fn)
-      (meth, targs, Nil)
+      val (meth, targss, args) = decomposeCall(fn)
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/ast/TreeInfo.scala]: I don't understand the change. Can you give an example where this matters?

File: compiler/src/dotty/tools/dotc/typer/Applications.scala [hunk 2]
```diff
@@ -962,9 +962,10 @@ trait Applications extends Compatibility { self: Typer with Dynamic =>
 
         val dummyArg = dummyTreeOfType(ownType)
         val unapplyApp = typedExpr(untpd.TypedSplice(Apply(unapplyFn, dummyArg :: Nil)))
-        val unapplyImplicits = unapplyApp match {
+        def unapplyImplicits(unapp: Tree = unapplyApp): List[Tree] = unapp match {
           case Apply(Apply(unapply, `dummyArg` :: Nil), args2) => assert(args2.nonEmpty); args2
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/typer/Applications.scala]: I think it's safer in that case to not have a default argument for `unapp`.

File: compiler/src/dotty/tools/dotc/typer/ReTyper.scala [hunk 3]
```diff
@@ -48,11 +52,22 @@ class ReTyper extends Typer {
   override def typedSuper(tree: untpd.Super, pt: Type)(implicit ctx: Context): Tree =
     promote(tree)
 
+  override def typedTyped(tree: untpd.Typed, pt: Type)(implicit ctx: Context): Tree = {
+    assertTyped(tree)
+    val type1 = checkSimpleKinded(typedType(tree.tpt))
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/typer/ReTyper.scala]: Rename `type1 -> tpt1`, `tree1 -> expr1` to conform to naming conventions used elsewhere. (`tpt` stands for "type tree")

File: compiler/src/dotty/tools/dotc/ast/TreeInfo.scala (hunk 1 above)
Comment (OlivierBlanvillain) [reply compiler/src/dotty/tools/dotc/ast/TreeInfo.scala]: The test added in the same commit (e199fc838e8c437c6cf88b7f27bbb6f350ed3bb2) breaks with a match error before this change

**Comment by OlivierBlanvillain** [comment]
Completely forgot about this PR :sweat_smile:. I've addressed the review and rebased, so it should be good to go!

File: tests/pos/inline-t9232b.scala [hunk 4]
```diff
@@ -0,0 +1,23 @@
+final class Foo(val value: Int)
+
+object Foo {
+  inline def unapplySeq(foo: Foo): Some[Seq[Int]] = Some(List(foo.value))
```
Comment (allanrenucci) [inline tests/pos/inline-t9232b.scala]: Should this be changed to `inline def unapply(foo: Foo)` after #3747?

Comment (OlivierBlanvillain) [reply tests/pos/inline-t9232b.scala]: I would say not for now, if/when `unapplySeq` is effectively not supported in Dotty then we should turn this tests into a scala2 test.

**Pull request merged**